RFC-0003-arrow-flight-connector.md

```diff
@@ -0,0 +1,122 @@
+
+Flight descriptor will be created from a command of byte array. The concrete implementation is responsible for providing the byte array that differs from server to server. Flight Info will be got from the Flight Descriptor. Presto can create connector splits for each ticket from Flight Info so that data can be fetched in parallel by different worker nodes.
+
+![Connector splits in Presto](arrow-flight-connector/Connector-splits-in-Presto.png)
+
+#### Class diagram
+
+![Class diagram](arrow-flight-connector/Class-diagram.png)
+
+
+## Prestissimo Implementation
+A single Prestissimo connector will handle all concrete instances of the Java Arrow Flight connector. This is possible because the prestissimo connector does not have to deal with metadata fetch and discovery. Each split wraps a ticket and a list of locations, which is all[^1] that is required to fetch data from the Flight Service in an implementation-agnostic way.
+
+[^1]: Authentication is one exception which cannot be handled in a truly generic manner. A solution is proposed below.
+
+### Arrow Record Batch to Velox vectors
+Arrow arrays and Velox vectors binary compatible with the Arrow C Data Interface.
+- Velox provides bridge methods to convert Velox vectors to-and-from from the C Interface [(Link)](https://github.com/facebookincubator/velox/blob/main/velox/vector/arrow/Bridge.h)
+- The Arrow SDK provides bridge method to convert Arrow arrays to-and-from the C Interface [(Link)](https://arrow.apache.org/docs/cpp/api/c_abi.html)
+
+The Arrow Flight SDK returns data in the form of Arrow Arrays, which can be converted into Velox vectors by using the provided bridge methods, using the Arrow C Interface as an intermediate representation. In most cases, these are cheap operations that don't involve copying or translating the actual data buffers.
+
+### Plugin-style authentication
+The worker needs to authenticate with the Flight server to redeem the Flight Ticket. This cannot be handled in a truly generic manner since the Arrow Flight protocol allows a lot of freedom to implement custom authentication strategies. Each implementaion of a flight service may require custom code to handle authentication.
+
+We propose to solve this by adding plugin-style authenticators to the prestissimo connectors. A Flight catalog can optionally specify an authenticator to use:
+```properties
+connector.name=arrow-flight
+flight.auth.name=simple-user-pass
+
+# The following are examples of authenticator-specific properties
+# In presto, the concrete module should handle them
+# In prestissimo, the authenticator plugin will handle them
+flight.auth.user=me
+flight.auth.password=itsme
+```
+In the above case, the authenticator registered as `simple-user-pass` will be used to authenticate all Flight requests for that catalog. More authenticators can be created and registered using an interface similar to the interface for connector creation and registration in Velox.
```

Comment:
Sounds like these authenticators should be registered in PrestoServer class. Is the AuthenticatorClass a pure Velox interface independent of Flight ? Wanted to check if there are flight dependencies for it... If there is then we will need to add dependencies in Velox.

Reply:
For convenience's sake, the authenticator is a Flight-dependent interface. I'm imagining that it would only be used alongside the arrow flight connector.

The interface itself can certainly be made independent of the flight connector; the tradeoff is that authenticator implementations that need the Flight SDK may end up having to write a few extra lines of boilerplate.

Should we make the interface independent?

Reply:
@ashkrisk : Its better to make the interface independent. And have customizations for flight.

Reply:
I'm not sure if there is much value in making the interface independent if it's not getting used elsewhere. Perhaps we can resolve this during code review once you've had a chance to see what the current implementation looks like and if you feel it's the sort of thing that you might want to use outside the Flight connector?

Reply:
@ashkrisk : My main concern with adding Flight related code in the generic interface is because flight libraries would compile with the velox core. We try to avoid dependent libraries as much as possible and limit them to only when the flight connector is needed.

Reply:
@aditi-pandit the Authenticator interface is defined within the folder for the Arrow Flight connector and is only enabled/compiled when Prestissimo is built with `PRESTO_ENABLE_ARROW_FLIGHT_CONNECTOR=ON`. Authenticator implementations would have a dependency on the Arrow Flight libraries, but if you don't build the connector (and hence don't build any authenticators for it) you don't need the Flight libraries at all.

Does this resolve your concern?

Reply:
@ashkrisk : This sounds a bit odd. Though we can look at your code to resolve this.